### PR TITLE
Update list of banned url shorteners in rake task

### LINKED ISSUE #1261: Regularly update the list of URL shorteners
Lobsters blocks URL shorteners like t.co and bit.ly because they create dupe posts and generate good traffic data for marketers. This used to be a hardcoded constant called `TRACKING_DOMAINS` that was irregularly updated; that was folded into the general-purpose domain banning feature in #1080.

I saw a story about frustrations [growing a link shortener](https://app.y.gy/blog/honeypot) ([via](https://news.ycombinator.com/item?id=39564678)) and the author mentioned [maintaining a list of shorteners](https://github.com/mayakyler/link-shorteners).

I'd like to use this list (and/or another, if others exist?) to replace our use of the domain banning feature, which requires manual intervention, doesn't provide a convenient list of blocked shorteners, and has the slightly odd step of creating a Domain record we haven't ever had a link to in order to ban it.

Updating our copy of the list could added to our daily cron job.

### PR DESCRIPTION
**Pull request by kernal053**
Resolve #1261

### CODE REVIEW

**Comment by pushcx** [comment]
Thanks for taking a look at this. That build failure is unrelated, it’s some database fixes making a test with generated data flaky. I’ll handle it.

File: lib/tasks/update_banned_url_shorteners.rake [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  url = "https://raw.githubusercontent.com/mayakyler/link-shorteners/refs/heads/main/js-link-shorteners/src/link-shorteners.txt"
+
+  begin
+    response = URI.parse(url).open
+    content = response.read
+
+    domains = content.split("\n").reject { |line| line.empty? || line.starts_with?("#") }
+
+    banned_by_user = User.find_by(username: "pushcx")
+    existing_domains = Domain.pluck(:domain)
```
Comment (pushcx) [inline lib/tasks/update_banned_url_shorteners.rake:17]: Please filter to only banned domains. There’s no need to pull all 30k records.

File: lib/tasks/update_banned_url_shorteners.rake [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+task update_banned_url_shorteners: :environment do
+  url = "https://raw.githubusercontent.com/mayakyler/link-shorteners/refs/heads/main/js-link-shorteners/src/link-shorteners.txt"
+
+  begin
+    response = URI.parse(url).open
+    content = response.read
+
+    domains = content.split("\n").reject { |line| line.empty? || line.starts_with?("#") }
+
+    banned_by_user = User.find_by(username: "pushcx")
```
Comment (pushcx) [inline lib/tasks/update_banned_url_shorteners.rake:16]: Hmm. This isn’t going to work for sister sites using our codebase and I don’t have a better idea for it. Maybe a new setting in the site-wide settings at the end of `config/application.rb`?

File: lib/tasks/update_banned_url_shorteners.rake [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    existing_domains = Domain.pluck(:domain)
+    new_domains = domains - existing_domains
+
+    if new_domains.any?
+      new_domains.each do |new_domain|
+        domain = Domain.find_or_create_by!(domain: new_domain)
+        domain.ban_by_user_for_reason!(banned_by_user, "Used for link shortening and ad tracking")
```
Comment (pushcx) [inline lib/tasks/update_banned_url_shorteners.rake:23]:
```suggestion
        domain.ban_by_user_for_reason!(banned_by_user, "Used for link shortening and ad tracking, see https://github.com/mayakyler/link-shorteners")
```

**Review by pushcx — CHANGES_REQUESTED** [review]
One puzzle, couple small changes.

File: lib/tasks/update_banned_url_shorteners.rake [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+desc "Update banned list of URL shorteners"
+
+task update_banned_url_shorteners: :environment do
+  url = "https://raw.githubusercontent.com/mayakyler/link-shorteners/refs/heads/main/js-link-shorteners/src/link-shorteners.txt"
+
+  begin
+    response = URI.parse(url).open
+    content = response.read
+
+    domains = content.split("\n").reject { |line| line.empty? || line.starts_with?("#") }
```
Comment (pushcx) [inline lib/tasks/update_banned_url_shorteners.rake:14]: Why did you write this `reject` filter?

**Review by pushcx — CHANGES_REQUESTED** [review]
OK, sounds good.

Please give this a run in your dev instance to handle the couple bugs and I look forward to merging this.

File: lib/tasks/update_banned_url_shorteners.rake (hunk 4 above)
Comment (kernal053) [reply lib/tasks/update_banned_url_shorteners.rake:14]: Oops. I was checking Peter Dave's Link Shortener list. Had to used it there. Removing it.

**Comment by kernal053** [comment]
Successfully testes this on my dev instance.

**Comment by pushcx** [comment]
Ah, I was confused, I'm sorry. I assumed some layer of our URL matching or Domain model would prohibit saving a Domain model with a `domain` attribute including non-ascii characters (or at least would force it to be encoded in punycode). There are a couple at the very end of the shortener list. I just checked the production database and, nope, I misremembered. We do have a handful of stories with domains that use non-latin1 characters, so it's clearly permitted.

I'm going to go ahead and merge this. I don't want to blow up the moderation log with ~2100 entries for new shorteners so I'll keep the bug open to add the task of writing a migration to block them 100 at a time so there are only ~21 log entries.

**Review by pushcx — APPROVED** [review]

**Pull request merged**